[FEATURE] Mettre à jour le prénom et le nom reçus du GAR quand le samlID change pour un même user (PIX-4771)

lego-technix opened the pull request:
## :unicorn: Problème

Le prénom et le nom reçus du GAR ne sont pas mis à jour quand le samlID change pour un même user.

## :robot: Proposition

Mettre à jour le prénom et le nom reçus du GAR quand le samlID change pour un même user.

## :rainbow: Remarques

RAS

## :100: Pour tester

On ne peut pas tester cette PR en RA.

Pour tester la PR il faut tester en local avec un _faux GAR_ à installer en local : 
https://github.com/1024pix/pix-saml-idp

En recette on pourra tester avec le _faux GAR_ suivant : 
https://test-idp.integration.pix.fr/

---

[fichiers-import-siecle.zip](https://github.com/1024pix/pix/files/12196004/fichiers-import-siecle.zip) d'exemple pour faire les tests ci-dessous, à utiliser pour faire les imports dans Pix Orga suivants les consignes ci-dessous.

---

1. Créer ou s'assurer que dans une organisation il existe un _organization learner_ avec un nom, prénom et date de naissance (par exemple `Juliette`, `Dumoulin`, `11/11/2011`). On pourra utiliser le fichier `siecle-julie-1.xml` du fichier ZIP en pièce jointe de ce ticket.
2. En utilisant un _faux GAR_, se générer un faux token d'authentification avec les mêmes nom et prénom que ceux de l'_organization learner_ et avec un _Saml ID_ (qui est le `externalIdentifier`) ayant une valeur particulière au choix (par exemple `abc123`)
3. Après la redirection automatique effectuée par le _faux GAR_ (_« Commencez votre parcours Pix, 
Inscrivez-vous ou connectez-vous sur la plateforme Pix et lancez votre test, Je commence »_), participer à une campagne de cette organisation en entrant le code de campagne associé (par exemple `SCOBADGE1`)
4. Faire une réconciliation avec les mêmes nom, prénom (par exemple `Juliette`, `Dumoulin`) et date de naissance (par exemple  `11/11/2011`) que ceux de l'_organization learner_
5. Constater que les champs `authentication-methods.externalIdentifier` et `authentication-methods.authenticationComplement` ont respectivement les valeurs suivantes :

   ```
   abc123
   ```

   ```json
   {"firstName": "Juliette", "lastName": "Dumoulin"}
   ```
6. Modifier dans l'organisation le nom, prénom de l'_organization learner_ pour leur donner de nouvelles valeurs (par exemple `Julie`, `Dumoulin-Lemarchand`). On pourra utiliser le fichier `siecle-julie-2.xml` du fichier ZIP en pièce jointe de ce ticket.
7. En utilisant un faux GAR, se générer à nouveau un faux token d'authentification avec les mêmes valeurs pour le nom et le prénom que les nouvelles valeurs de ceux de l'_organization learner_ (par exemple `Julie`, `Dumoulin-Lemarchand`) et avec un _Saml ID_ (qui est le `externalIdentifier`) ayant une nouvelle valeur particulière au choix (par exemple `xyz789`)
8. Après la redirection automatique effectuée par le _faux GAR_ (_« Commencez votre parcours Pix, 
Inscrivez-vous ou connectez-vous sur la plateforme Pix et lancez votre test, Je commence »_), participer à nouveau à une campagne de cette organisation en entrant le code de campagne associé (par exemple `SCOBADGE1`)
9. Faire à nouveau une réconciliation avec les nouvelles valeurs des nom et prénom de l'_organization learner_ (par exemple `Julie`, `Dumoulin-Lemarchand`) et toujours la même date de naissance  (par exemple  `11/11/2011`)
10. Constater que les champs `authentication-methods.externalIdentifier` et `authentication-methods.authenticationComplement` ont été mis à jour avec respectivement les valeurs suivantes :
   ```
   xyz789
   ```
   
   ```json
   {"firstName": "Julie", "lastName": "Dumoulin-Lemarchand"}
   ```

lego-technix committed:
refactor(api): move non-exported functions at the end of the module + reorder exports order

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [App (.fr)](https://app-pr6717.review.pix.fr)
  * [App (.org)](https://app-pr6717.review.pix.org)
  * [Orga (.fr)](https://orga-pr6717.review.pix.fr)
  * [Orga (.org)](https://orga-pr6717.review.pix.org)
  * [Certif (.fr)](https://certif-pr6717.review.pix.fr)
  * [Certif (.org)](https://certif-pr6717.review.pix.org)
  * [Admin](https://admin-pr6717.review.pix.fr)
  * [API](https://api-pr6717.review.pix.fr/api/)
  * [Pix1D](https://1d-pr6717.review.pix.fr)

Les variables d'environnement seront accessibles via les liens suivants :
  * [scalingo front](https://dashboard.scalingo.com/apps/osc-fr1/pix-front-review-pr6717/environment)
  * [scalingo api](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr6717/environment)

lego-technix committed:
refactor(api): put const code at the root scope of the module + more explicit name

lego-technix committed:
test(api): normalize and improve test texts

lego-technix committed:
refactor(api): rename checkIfStudentHasAnAlreadyReconciledAccount

into assertStudentHasAnAlreadyReconciledAccount because this method does not
return any value but throws instead when assertion is not met.

lego-technix committed:
refactor(api): add TODO

lego-technix committed:
refactor(api): improve error messages in user-reconciliation-service

lego-technix committed:
test(api): add a test context

lego-technix committed:
test(api): add more assertions

lego-technix committed:
refactor(api): some factoring-out of code

lego-technix committed:
feat(api): update firstName and lastName too when reconciling and updating the samlId

lego-technix committed:
feat(api): add transaction in createUserAndReconcileToOrganizationLearnerFromExternalUser

Co-authored-by: Ismael Gorissen <ismael.gorissen@gmail.com>
Co-authored-by: BKsensei <rebecca.kaci@gmail.com>
Co-authored-by: Thomas Evano <thomas.evano@pix.fr>

igorissen commented on api/lib/domain/usecases/create-user-and-reconcile-to-organization-learner-from-external-user.js:
```diff
@@ -24,40 +31,34 @@ const createUserAndReconcileToOrganizationLearnerFromExternalUser = async functi
 
   try {
     matchedOrganizationLearner =
+      // TODO: Rename findMatchingOrganizationLearnerIdForGivenOrganizationIdAndUser→findMatchingOrganizationLearnerForGivenOrganizationIdAndReconciliationInfo
       await userReconciliationService.findMatchingOrganizationLearnerIdForGivenOrganizationIdAndUser({
```
```suggestion
      await userReconciliationService.findMatchingOrganizationLearnerForGivenOrganizationIdAndReconciliationInfo({
```

lego-technix committed:
refactor(api): rename findMatchingOrganizationLearnerIdForGivenOrganizationIdAndUser

into findMatchingOrganizationLearnerForGivenOrganizationIdAndReconciliationInfo
to state what the function really do.

Co-authored-by: Ismael Gorissen <ismael.gorissen@gmail.com>

igorissen reviewed (approved):
Tests fonctionnels ✅

Thomasevano reviewed (approved):
Testé fonctionnellement ✅ 
J'ai du créer l'`organization-learner` manuellement en bdd
